CMake/build.cmake

```diff
@@ -13,23 +15,37 @@ if(NOT CMAKE_BUILD_TYPE)
   set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
 endif()
 
+# SENSEI should always be position independent
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+# SENSEI requires minimum C++11
+if (NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 11)
+  set(CMAKE_CXX_STANDARD 11)
+endif ()
+
 if (NOT MSVC)
-  if (NOT CMAKE_CXX_FLAGS)
-  set(tmp "-fPIC -std=c++11 -Wall -Wextra")
+  string(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra")
+
+  # Need to explicitly set optimization level to -O3 because some systems default to -O2
+  set(CMAKE_CXX_FLAGS_RELEASE "-O3 -DNDEBUG")
+  set(optimization_flags " -mtune=native; -march=native")
+  foreach (flag ${optimization_flags})
+    check_cxx_compiler_flag(${flag} has_flag)
+    if(has_flag)
+      string(APPEND CMAKE_CXX_FLAGS_RELEASE ${flag})
+    endif ()
+  endforeach ()
+
   if (BUILD_STATIC_EXECS)
-    set(tmp "${tmp} -static -static-libgcc -static-libstdc++ -pthread -Wl,-Bstatic")
-  endif()
-  if ((APPLE) AND ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang"))
-    set(tmp "${tmp} -stdlib=libc++")
-  endif()
-  if ("${CMAKE_BUILD_TYPE}" MATCHES "Release")
-    set(tmp "${tmp} -O3 -march=native -mtune=native")
-  endif()
-  set(CMAKE_CXX_FLAGS "${tmp}"
-    CACHE STRING "SENSEI build defaults"
-    FORCE)
-  endif()
-endif()
+    string(APPEND CMAKE_CXX_FLAGS "-static -static-libgcc -static-libstdc++ -pthread -Wl,-Bstatic")
```

Comment:
This only works for GCC

Reply:
You are very likely correct. However, this is for our own use so that we can reproduce the results of our sc16 paper which made use of static linked executables (SENSEI's miniapps). I suggest we leave this option so we do not lose that information. Those who want to use the feature will do so at their own peril, and could submit a patch to support other compilers if it is desired.

Reply:
note: these were so that we can reproduce the results from the sc16 paper. others will use them at their own peril.

Reply:
Since this is behind an option that is default off, I am going to call this resolved for now.